internal/file/file.go

```diff
@@ -0,0 +1,101 @@
+package file
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/spf13/afero"
+)
+
+const hashExt = ".sha256"
+
+// Copy copies a file from the given bytes to destination.
+func Copy(destination string, fin []byte) error {
+	if _, err := os.Stat(destination); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("could not stat file: %w", err)
+	}
+
+	fout, err := os.Create(destination)
+	if err != nil {
+		return fmt.Errorf("could not create file: %w", err)
+	}
+	defer fout.Close()
+
+	r := bytes.NewReader(fin)
+
+	if _, err := io.Copy(fout, r); err != nil {
+		return fmt.Errorf("could copy file: %w", err)
+	}
+
+	return nil
+}
+
+// CheckSize checks if a file exists and has an exact size.
+func CheckSize(filename string, size int) (bool, error) {
+	stat, err := os.Stat(filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("could not stat file: %w", err)
+	}
+	return stat.Size() == int64(size), nil
+}
+
+var testableFS = afero.NewOsFs()
+
+// CheckHash checks if a file has the given SHA256 hash.
+// It supports reading the file's current hash from a static file saved next to it with the hashExt extension.
+func CheckHash(filename, hash string) (bool, error) {
+	if fh, err := afero.ReadFile(testableFS, filename+hashExt); err == nil {
+		return string(fh) == hash, nil
+	}
+	fh, err := sha256Sum(testableFS, filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return true, err
+	}
+	return fh == hash, nil
+}
```

Comment:
The hash should be computed from the file itself, otherwise the hash can be tampered in the same way that the file was tampered.

Reply:
It's not used for tamper protection as you say - the pregenerated hashes could almost just be random numbers - I went for real hashes just so there's the possibility of checking against accidental corruption of the copied files themselves and the fallback of being able to recompute when the hash file wasn't copied. But broadly I was going for not having to compute hashes at runtime at all.

Reply:
Recomputing the hash from the file on every run would theoretically use time and memory that we don't want or need to use - the purpose of this PR is to make sure we have updated files when _we_ change them, the user is free to break their own installation (and it can be fixed again by deleting the main file or hash).

Reply:
I'm reworking this to avoid saving hashes, except on Windows where it seems it's needed (because the zip is expanded)